analyzers/src/SonarAnalyzer.Common/Rules/HardcodedIpAddressBase.cs

```diff
@@ -20,74 +20,130 @@
+        private const int NumOfIPv4AddressParts = 4;
+
+        private readonly ISet<string> ignoredVariableNames =
             new HashSet<string>
             {
                 "VERSION",
                 "ASSEMBLY",
             };
 
+        private readonly Regex ipv6LoopbackPattern = new Regex("^(?:0*:)*?:?0*1$", RegexOptions.Compiled);
+
+        private readonly Regex ipv6NonRoutablePattern = new Regex("^(?:0*:)*?:?0*$", RegexOptions.Compiled);
+
+        private ImmutableArray<DiagnosticDescriptor> supportedDiagnostics;
+
+        protected abstract GeneratedCodeRecognizer GeneratedCodeRecognizer { get; }
+
+        protected abstract TSyntaxKind SyntaxKind { get; }
+
+        protected abstract DiagnosticDescriptor Rule { get; }
+
         protected abstract string GetAssignedVariableName(TLiteralExpression stringLiteral);
+
         protected abstract string GetValueText(TLiteralExpression literalExpression);
+
         protected abstract bool HasAttributes(TLiteralExpression literalExpression);
 
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+        {
+            get
+            {
+                if (supportedDiagnostics == null)
+                {
+                    supportedDiagnostics = ImmutableArray.Create(Rule);
+                }
+
+                return supportedDiagnostics;
+            }
+        }
+
         protected HardcodedIpAddressBase(IAnalyzerConfiguration analyzerConfiguration) : base(analyzerConfiguration)
         {
         }
 
-        protected Action<SyntaxNodeAnalysisContext> GetAnalysisAction(DiagnosticDescriptor rule) =>
-            c =>
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(GeneratedCodeRecognizer, AnalyzeLiteral, SyntaxKind);
+
+        private void AnalyzeLiteral(SyntaxNodeAnalysisContext context)
+        {
+            if (!IsEnabled(context.Options))
             {
-                var stringLiteral = (TLiteralExpression)c.Node;
-                var literalValue = GetValueText(stringLiteral);
-                if (IsLocalHost(literalValue))
-                {
-                    return;
-                }
+                return;
+            }
 
-                var variableName = GetAssignedVariableName(stringLiteral);
-                if (variableName != null &&
-                    IgnoredVariableNames.Any(variableName.Contains))
-                {
-                    return;
-                }
+            CheckForHardcodedIpAddresses(context);
+        }
 
-                if (HasAttributes(stringLiteral))
-                {
-                    return;
-                }
+        private void CheckForHardcodedIpAddresses(SyntaxNodeAnalysisContext context)
+        {
+            var stringLiteral = (TLiteralExpression)context.Node;
+            var variableName = GetAssignedVariableName(stringLiteral);
```

Comment:
Fixing old technical debt: 
* `ignoredVariableNames` contains upper case values
* `Contains` below is case sensitive
* There's no sign that `GetAssignedVariableName` should or does return upper case
* Inheriting classes do `ToUpperInvariant`
* These facts are hard to follow

Let's 
* Add `StringComparer.InvariantCultureIgnoreCase` to that `HashSet` constructor argument to have case insensitive matching
* Remove `ToUpperInvariant` from C# and VB